Develop

masa-s60 opened the pull request:
データベースのデータ操作やログイン管理に関する改善点があればアドバイス宜しくお願い致します。

masa-s60 committed:
[add]firebaseを導入し、制作を開始

masa-s60 committed:
[add]仮コミット

masa-s60 committed:
[add]仮コミット

masa-s60 committed:
[add]実装完了

masa-s60 commented:
・ログイン状態の保持について
今回の私の実装は、ログインしている状態から再読み込みボタン（どのブラウザにもあるwebページの画面左上にある丸い矢印）を押下するとログアウト状態になり、再度ログインしなければならない仕様となってしまっております。

課題の「ログアウトされた状態でダッシュボードへアクセスすると、ログイン画面へリダイレクト」の要件は満たしているので課題は提出しておりますが、通常のログイン機能ではこのような事象は起きてはならないと思っています。

実装を試みるため「react 画面更新時 コンテキスト保持」等で検索しても、実装済みのuseContext情報ばかりで、再読み込みしてもコンテキストを保持できる方法が見つかりませんでした。

何かいい方法があればご教示お願いしたいです。

・実装出来ている関連機能
ログイン後に「戻る」ボタンを押下してもログイン画面には戻らない
ログアウト状態から「進む」ボタンを押下しても投げ銭アプリ画面には進まない
ログアウトボタン押下でログアウトする

masa-s60 commented:
・モーダルウィンドウのコンポーネントについて
実装していてモーダルウィンドウのコンポーネントをどこに配置するのか、という問題が非常に悩みました。

今回の実装では、ユーザ一覧のコンポーネント「user-table.tsx」でインポートして取り扱っております。
出来れば余計なuseStateのprop経由を避けるため、「walletをみる」「送る」ボタンがあるコンポーネント「user-list.tsx」に置きたかったのですが、tableタグにはtrタグやtdタグ以外のものは含めてはいけないとエラーになってしまったので、仕方なく上階層のコンポーネント「user-table.tsx」のtableタグ外に配置しました。

reactではモーダルウィンドウはコンポーネントとして切り分けず直に記述する、というような記事も見ましたが、
結局何が一番いいのかわからなくなりました。

こちらも適切な実装方法があれば教えて頂きたいです。

masa-s60 committed:
[remove]不要ファイルの削除

masa-s60 committed:
[add,remove]画像ファイル追加と不要コードの削除
react画像ファイル削除後にエラーが発生したため戻しました。不要ログ出力を削除しました。

masa-s60 committed:
[remove]不要コードの削除

yoshin04 commented:
> 今回の私の実装は、ログインしている状態から再読み込みボタン（どのブラウザにもあるwebページの画面左上にある丸い矢印）を押下するとログアウト状態になり、再度ログインしなければならない仕様となってしまっております。
実装を試みるため「react 画面更新時 コンテキスト保持」等で検索しても、実装済みのuseContext情報ばかりで、再読み込みしてもコンテキストを保持できる方法が見つかりませんでした。

何かいい方法があればご教示お願いしたいです


やり方は色々あるのですが、状態管理ライブラリを使ってsessionストレージ、もしくはローカルストレージに保持するのが良いかなと思います。
next.jsを使用している場合は、next-authを使ったりもするのですが、状態管理ライブラリで実装する場合はrecoilを使ったりしますね！


> 今回の実装では、ユーザ一覧のコンポーネント「user-table.tsx」でインポートして取り扱っております。
出来れば余計なuseStateのprop経由を避けるため、「walletをみる」「送る」ボタンがあるコンポーネント「user-list.tsx」に置きたかったのですが、tableタグにはtrタグやtdタグ以外のものは含めてはいけないとエラーになってしまったので、仕方なく上階層のコンポーネント「user-table.tsx」のtableタグ外に配置しました。

今回の実装で個人的には良いと思います！
割と僕はpropsのバケツリレーは悪くないという考えです。
そのコンポーネントに対して適切なスコープのpropsを定義するほうがコードが読みやすいと思います。

逆にpropsのバケツリレーを避けてすべてグローバルステートとかにしてしまうと、すべてグローバル変数で定義しているのと変わらないので、確認しなきゃいけないところとコードの変更による見えない影響範囲がでかくなってしまうと考えている感じです！

yoshin04 commented on src/components/Common/Molecules/user-list.tsx:
```diff
@@ -0,0 +1,50 @@
+import { useStateTypeShow, useStateTypeUserInfo, useStateTypeUsers, typeSetActive, typeUser } from "../../../types/type";
+import { useAuthContext } from "../../../Context/auth-context";
+
+const UserList: FC<{userInfoItem: useStateTypeUserInfo, showModalItem: useStateTypeShow, usersItem: useStateTypeUsers, setActive: typeSetActive}> = (props) => {
```
こちらのコンポーネントはOrgainismsになりますね。
分け方のコツは明確に意味を持つコンポーネントは、Organismsに該当する可能性が高いです。

今回でいうとユーザーのリストという意味を持っていて、汎用性がユーザーのリストに限定されるためOrganismsです。

masa-s60 replied:
Orgainismsの存在が抜けていました。

今一度コンポーネント分割について理解を深めたいと思うのですが、前回の実装でアドバイス頂いたFeaturesについて少し疑問を感じました。

色々と調べましたが、このFeaturesはゼネラル・エレクトリック社が考案したコンポーネントの分割単位で、アトミックデザインから派生した考え方、という事でした。
コンポーネントの考え方には各々が捉えやすいようにいろんな表現を試みている、という事も知りました。

以上を踏まえて私が感じたのは、アトミックデザインの中にFeaturesを共存させない方がいいのではないでしょうか。

ゼネラル・エレクトリック社はコンポーネントを
「Principles、Basics、Components、Templates、Features、Applications」
として分割し、その中で扱っているものなので、この一部であるFeaturesをアトミックデザインに取り入れると、考え方が変わったり、分割方法に悩む結果になってしまうのではないかと思いました。

なので、現在私がFeaturesに置いているコンポーネントを全てOrgainismsに変更しようと思うのですが、今の実装でFeaturesを置いて何かメリットがあるとすれば、どんな考え方なのでしょうか。

玄人の方々の視点で考えるコンポーネント分割についてもう少しお聞きできればと思います。

yoshin04 replied:
> 以上を踏まえて私が感じたのは、アトミックデザインの中にFeaturesを共存させない方がいいのではないでしょうか。

これは、プロジェクトやプロジェクトの状態によりますね。
例えば、プロジェクトが小さいうちはatomicデザインで管理していても良いですが、organismsが大きくなりすぎて扱いにくくなります。
featureを用意すると例えばuser関連でしか使わないという形で示せるので、管理が楽になります。

masa-s60 replied:
プロジェクトが大きくなるにつれてorganismsも増えてくるという事ですね。
であれば、大きなorganismsはatomicデザインでいうTemplatesとして扱えばいいのではないかと思ってしまいますが。

感覚で言うと、featureを導入するとなるとfeature立ち位置は
Molecules < Feature < Organisms
といったところでしょうか。

一先ず今回の自分のコンポーネント分割はこの感覚で行っています。

yoshin04 replied:
大きなOrganismsというのが、複数のOrganismsの集合体を想像しているのであれば、templateとして定義するのでもありです。

>感覚で言うと、featureを導入するとなるとfeature立ち位置は
Molecules < Feature < Organisms
といったところでしょうか。

これの意味がわからずでして、「<」記号はxxより下を指すので、ぱっと見コンポーネントを定義する優先度のように見えます。
Featureのなかには必ずOrganismsが入るわけではないです。
usersというFeatureであれば、usersに関係するコンポーネントになります。
moleculesは基本どのページでも使えるようなコンポーネントですので、Featureの中に定義することはあまりないです。
逆にOrganismsでも汎用的に使えるものはFeatureの中には入らないです。
例えば、UserListコンポーネント(ユーザーのリストを表示するコンポーネント)などといったUser関連でしか使わないコンポーネントが該当します。

masa-s60 replied:
Molecules < Feature < Organismsは、括りの大きさとして記載しました。
FeatureはOrganismsやMoleculesの集まりで、Organismsほど大きな集合体ではない、
という位置づけなのかという質問でした。

これまでのFeatureに関するご説明と調べた結果を要約すると、以下の通りでしょうか。

・Organismsよりも用途が特定される
・「Moleculesよりも大きい」「Templatesよりも小さい」といった概念ではない
・Featureディレクトリはcomponents内に限定されず、src直下などでも機能の括りとしてFeatureを置くことがある
・「login-form.tsx」「sign-up-form.tsx」「send-money-form.tsx」は、それぞれフォーム機能の用途が限定されているためFeatureとなる
・「user-table.tsx」は、Organismsの「user-list.tsx」を含んでおり、且つuser関連では汎用的に扱えるため、OrganismsもしくはTemplateである

間違いがあればご指摘宜しくお願い致します。

masa-s60 replied:
まだアトミックデザインの考え方が誤っている可能性があるので、誤認を正すためにも改めて確認したいです。
過度な誤認の場合、大幅なコーディング修正となってしまうので、先に質問させていただきます。

============== 以下、参考サイトによるTemplatesの説明（複数引用）================

・“テンプレート”とは、画面のレイアウトのことです。原子と分子、生体を組み合わせて作成します。
この段階では、まだ具体的なテキストや画像は挿入されません。

・原子、分子、有機体をまとめてWebサイトの基礎となるものがテンプレートです。
一般的にイメージされるWebサイトのテンプレートと同様に、これを用意するだけで基本的なWebサイトの機能やデザインが完成しています。

・原子や分子、有機体をさらに組み合わせて全体の構成を作り上げたものがテンプレートです。この時点では画像や詳細なテキストは流し込まず、ワイヤーフレームのような感覚で画面を作っていきます。
そのため、ここでは情報の過不足がないか、機能としての不足はないかといった観点で作成することが大切です。

==================================================================

上記の事から、
Templatesはテンプレートとして使い回せるような設計すべきで、具体的なテキスト等は全てTemplatesより大きな括りであるpagesから注入するのではないか、という理解に至りました。

AtomsとMoleculesについては、ご教示により理解できていると思っておりますが、
その他のアトミックデザインの構成において、以下の私の見解に誤りやアドバイスがあればご指摘宜しくお願い致します。

・具体的なテキストなどのpropsは全てPagesコンポーネントから渡す必要がある。つまりPages以外は全て汎用性が効くような実装になる。

・今回の私の実装のheader.tsx, footer.tsxは、AtomsやMoleculesのような小さい単位ではあるが、括りとしてはページ全体の構成するためのTemplatesにあたる。

・modal-window.tsxの位置付けは、TemplatesとしてもFeatureとしてもつじつまが合わなくなってしまうので、Pagesとする。

※つじつまが合わなくなる内容
1.modal-window.tsxをTemplatesとした場合、
modal-window.tsxはuse-table.tsxの中にあり、use-table.tsxはuse-list.tsx(Organisms)が入っているので、Organisms括るためuse-table.tsxは必然的にTemplatesになる。
TemplatesにTemplatesが入っているのは矛盾しているのでNG。

2.modal-window.tsxをFeatureとした場合、
modal-send.tsxやmodal-wallet.tsxはOrganismsとなり、modal-send.tsx内のsend-money-form.tsxはform要素を含んでいるためMoleculesにはできず、必然的にsend-money-form.tsxもOrganismsとするしかないので、
Organismsの中にOrganismsが入るのは矛盾しているのでNG。
他に思いつく方法としては、modal-send.tsxもmodal-wallet.tsxもsend-money-form.tsxもすべて限定的であるためFeatureとする、です。

TemplatesにTemplates、OrganismsにOrganisms、FeatureにFeatureが入っても良いのであれば特に問題ありません。

yoshin04 replied:
> ・Organismsよりも用途が特定される
・「Moleculesよりも大きい」「Templatesよりも小さい」といった概念ではない
・Featureディレクトリはcomponents内に限定されず、src直下などでも機能の括りとしてFeatureを置くことがある

はい。
そもそもFeaturesはアトミックデザインのようなコンポーネントを切り離す概念ではないです。
デザインではなく、コード全体をわかりやすく設計するために導入します。

Featuresというのはカテゴライズであり、Featureの中にmoleculesやOrganismsを配置したりします。
要は、Organismsの中でも特定の機能でしか使わないコンポーネント、汎用的に仕様することのできるコンポーネントがあります。
それらを一緒に管理して使うと大変だからFeaturesに切り出そうということです。
そのため、Featuresは単なるカテゴライズであり、コンポーネントだけが入るわけではないです。
hooksやexport させたい変数や関数など様々なファイルを格納します。

> ・「login-form.tsx」「sign-up-form.tsx」「send-money-form.tsx」は、それぞれフォーム機能の用途が限定されているためFeatureとなる

コンポーネントは、Organismsに当たります。
このOrganismsのコンポーネントの設計は、おっしゃる通りFeatureディレクトリの中にAuthを用意してそこに配置するのが良いと思います。

>・「user-table.tsx」は、Organismsの「user-list.tsx」を含んでおり、且つuser関連では汎用的に扱えるため、OrganismsもしくはTemplateである

テンプレートではないです。
テンプレートとは、限りになくページに近いです。
このコンポーネントは、テーブルやリストにあたるので、ページの1部品にしかすぎません。
そのため、Organismsに該当します。
そして、Featureディレクトリを用意しているのであれば、Featureディレクトリ内にusersを作り、そこにOrganimsのコンポーネントを配置します。
なぜならば、コンポーネント名にUserが含まれているということはUser関連の機能で仕様すると限定されているからです。

yoshin04 commented on src/components/Common/atoms/error-message.tsx:
```diff
@@ -0,0 +1,15 @@
+const ErrorMessage:FC<{errorMessage: string, errors: any}> = (props) => {
+  return(
+    <div>
+      {props.errorMessage && <p className="level-item has-text-danger mb-2">{props.errorMessage}</p>}
+      {props.errors.text?.message && <p className="level-item has-text-danger m-4">{props.errors.text?.message as string}</p>}
+      {props.errors.UserName?.message && <p className="level-item has-text-danger">{props.errors.UserName.message as string}</p>}
+      {props.errors.Email?.message && <p className="level-item has-text-danger">{props.errors.Email.message as string}</p>}
+      {props.errors.Password?.message && <p className="level-item has-text-danger">{props.errors.Password.message as string}</p>}
```
こう行く書き方はあまりやらないですね。
エラーのメッセージの出力が限定されているので、コンポーネントを分けるよりも必要な場所にエラーメッセージの出力のコードを書いたほうが良いです。

もしくは、childrenを使って汎用性を高くして、エラーメッセージのスタイルを適応するのみにするとかですね。
```
const ErrorText: FC<Props> = ({children}) => (
  <p className="text-red-5--"> {children} </p>
)
```
上記くらい汎用性が高いコンポーネントになるとatomsやmoleculesに該当します！

yoshin04 commented on src/components/Common/atoms/header-image.tsx:
```diff
@@ -0,0 +1,13 @@
+import { FC } from "react";
+
+const HeaderImage:FC<{iconValue: string}> = (props) => {
+  return(
+    <div className="level-item" style={{marginTop: '4vw',marginBottom: '10vw'}}>
+      <span className="icon">
+        <i className={props.iconValue} style={{fontSize: '12vw'}}></i>
+      </span>
+    </div>
+  )
+}
```
こちらはロゴのことですかね。
コンポーネント名からではヘッダーの画像で使うことが確定になってしまっていて汎用性がないため、atomsではないです。

IconImageとかにして、画像の大きさは親要素で管理できるようにする、もしくはpropsで大きさを親で管理できるようになるとatomsになります！

yoshin04 commented on src/components/Common/atoms/input-header.tsx:
```diff
@@ -0,0 +1,11 @@
+import { FC } from "react";
+
+const InputHeader: FC<{type: string}> = (props) => {
+  return(
+    <div>
+      <p className="input-header">{props.type}</p>
+    </div>
+  )
+}
```
InputHeaderとは、入力のラベルのことですかね。
であれば、labelタグでinputタグのところで定義するのが良いです。
例：
```
const TextInput: FC<props> = ({label, placeholder}) => (
  <div className="space-y-2>
    <label>{label}</label>
   <input type="text" placeholder={placeholder} />
  </div>
)
```

masa-s60 replied:
Inputコンポーネントにlabelを作成しましたが、こうなるとInputコンポーネントはアトミックデザイン的にはAtomsではなくMoleculesに該当するとような気もしますがどうでしょうか。

あくまでもMoleculesはAtomsコンポーネントの集まり、という解釈でしょうか。

yoshin04 replied:
ラベルまで、含めるのであればおっしゃる通りMoleculesで良いです！

> あくまでもMoleculesはAtomsコンポーネントの集まり、という解釈でしょうか。

はい。
ご認識の通り、Atomsの集まりになります。

yoshin04 commented on src/components/Common/atoms/input-tip.tsx:
```diff
@@ -0,0 +1,28 @@
+  return(
+    <div>
+      <input
+      {...register('text', { 
+        required: {
+          value: true,
+          message: '送る金額を入力してください',
+        },
+        maxLength: {
+          value: 8,
+          message: '入力桁数が多すぎます',
+        }
```
バリデーションの内容がここで決まってしまうので、react-hook-form のControllerを使って親要素でreact-hook-formの内容を管理できると、このコンポーネントの汎用性が高くなります。

masa-s60 replied:
Controllerを使用してご指摘通りの実装する事は出来ましたが、Controllerの参考ドキュメントが少なく知識が曖昧な部分があります。

参考ドキュメントはMaterial UIを使用した例ばかりで、通常のコンポーネントで同じようなやり方をしてもinputのテキスト内容がformに反映されなかったりで、うまくいきませんでした。

結局のところregisterをInputコンポーネントに渡したり、forwardRefを使用する事で解決できましたが、
Material UIの例にはregisterを渡す方法はどこにも記載されていませんでした。
Material UIはまだ使ったことがないのでお聞きしたいのですが、Material UIはコンポーネントの性質上registerは不要で、通常のコンポーネントではregisterが必要という事で良いのでしょうか。
また、forwardRef使用時は関数の引数にrefが必須でしたが、refをinputにあてるとバリデーションがうまくいかなかったので使用していません。
registerにControllerと同様の名をあてる事で期待通りに機能しました。

つまり、registerでrefの代わりにinputタグのHTMLElementを取得させる事で、Controllerにてバリデーション操作が可能になる、という事よろしいでしょうか。

yoshin04 replied:
> Material UIはまだ使ったことがないのでお聞きしたいのですが、Material UIはコンポーネントの性質上registerは不要で、通常のコンポーネントではregisterが必要という事で良いのでしょうか。

いえ、あまり、Material UIだからというのは関係ないですね。
結局のところ、自前のコンポーネントかUIコンポーネントにReact Hook Formコンポーネントを使うかの違いでしかないですね。
あと、Controllerを使うのであればregisterを使う必要はないですね。
既存Componentに対してReact Hook Formを使う場合はControllerを使う方が多いですね。
逆にHTML要素に直接という場合は、registerを使用するケースが多いです。

> Controllerを使用してご指摘通りの実装する事は出来ましたが、Controllerの参考ドキュメントが少なく知識が曖昧な部分があります。

個人的にはReact Hook Formのドキュメントや参考情報はかなり充実している方だと思っています。
例えば、Controllerの使い方の公式ドキュメントは以下のようにサンプルコード次で示しています。
https://www.react-hook-form.com/api/usecontroller/controller/

> つまり、registerでrefの代わりにinputタグのHTMLElementを取得させる事で、Controllerにてバリデーション操作が可能になる、という事よろしいでしょうか。

すみません。
こちら、私がライブラリを開発したわけではないのでちゃんとお答えできないです。
ただ、ライブラリを見る限りでは、ControllerのバリデーションにfowordRefの指定が必要とは書いていないです。
本来は、Controller内の記述を正しくしていれば、Controllerで設定したバリデーションは効くはずです。
そのため、ドキュメントの書き方に沿っていない可能性が高いです。
例えば、name属性の指定が間違っていないか、Controller使用時のバリデーションの書き方に誤りがないかを確認してみるのが良いと思います。

yoshin04 replied:
例えば、今実装しているコンポーネントを見るとControllerのfiledオブジェクト内にあるonChangeが使えていないです。
```
<Controller
                  name="Email"
                  control={control}
                  defaultValue=""
                  render={({ field }) => 
                    <Input
                      {...field}
                      type="text"
                      register={register}
                      label="メールアドレス"
                      classValueLabel="input-label"
                      classValueInput="input-style"
                      maxLength={50}
                      placeholder='Email'
                    />
                  }
                  rules={{ 
                    required: {
                      value: true,
                      message: "Emailを入力してください",
                    },
                    minLength: {
                      value: 8,
                      message: "Emailは8文字以上入力してください",
                    },
                    pattern: {
                      value: /.+@.+\..+/,
                      message: "メールアドレスを入力してください",
                    }
                  }}
                />
```

yoshin04 commented on src/components/Common/atoms/input-user-Info.tsx:
```diff
@@ -0,0 +1,27 @@
+            value: 8,
+            message: `${type}は8文字以上入力してください`,
+          },
+        })} 
```
バリデーションの内容がここで決まってしまうので、react-hook-form のControllerを使って親要素でreact-hook-formの内容を管理できると、このコンポーネントの汎用性が高くなります。

yoshin04 commented on src/components/Common/atoms/login-button.tsx:
```diff
@@ -0,0 +1,17 @@
+import { FC } from 'react';
+
+const LogInButton:FC<{buttonValue: string}> = (props) => {
+
+  return(
+    <div className="level-item">
+      <button 
+        className="mt-5 login-button-style"
+        type="submit" 
+      >
+        {props.buttonValue}
+      </button>
+    </div>
+  )
+}
```
ここもログインボタンとしてしまっているため、atomsにはならないです。
atomsにするのであれば、ボタンと幅広くするべきですね。
typeもtextかsubmitかをpropsで決められるようにして、button内のテキストはchildrenにする感じにするとatomsに該当するようになります！

masa-s60 replied:
[button内のテキストはchildrenにする感じにするとatomsに該当するようになります]

こちらを考慮すると、結局のところどちらがいいのかお聞きしたいです。

①Buttonコンポーネントで文字を囲ってchildrenとして渡す方法
`<Button>文字</Button>`

②Buttonコンポーネントにpropsとしてtextを渡す方法
`<Button text="文字" />`

子側の実装はほとんど変わらないので
`<button>{props.text} or {props.children}</button>`

親側の実装としては、今まで値はpropsで渡してきたぶん自分は②の方がいいと感じていますが、
①の方が都合がいい事があれば教えていただきたいです。

※指摘いただいたtitle.tsx等（現時点ではtext.tsx）におけるchildren運用の件も同様

yoshin04 replied:
>こちらを考慮すると、結局のところどちらがいいのかお聞きしたいです。

いえ、 ①の方が良いです。
なぜならば、1の方が汎用的に作れるからです。
```
<Button>
  <span style={}>
    text
  <span>
</Button>
```
実装の詳細によるのですが、childrenにしておけば、Button内のスタイルを変えたい場合(例えばpaddingを少し他のButtonよりも多めに取りたいなど)場合も対応できます。

yoshin04 commented on src/components/Common/atoms/logout-button.tsx:
```diff
@@ -0,0 +1,20 @@
+import { useAuthContext } from "../../../Context/auth-context";
+
+const LogOutButton = () => {
+  const context = useAuthContext();
+  const navigate = useNavigate();
+
+  const logout = () => {
+    context?.setUser(undefined);
+    navigate("/");
+  }
```
atomsは、汎用性があるべきですので基本的にはビジネスロジックを持たせては行けないです。
LoginFormなどいったOrganismsを定義して、そこからlogout処理などの具体的なビジネスロジックをatomsに対して注入するべきです。

masa-s60 replied:
このプロジェクトにおける全てのbutton要素をButtonコンポーネントで扱えるようまとめてみました。
汎用性を持たせるためにあえてしている事が2つあるのですが、それが適切がどうかを判断頂きたく存じます。

1.buttonタグのonClickで実行するビジネスロジックの関数はいろいろな用途の可能性があるため、ビジネスロジックに渡す引数に汎用性を持たせるため、引数の型をあえてanyにしました。

2.onClickなどのイベントはビジネスロジック関数を実行するため、ビジネスロジック関数がundefinedの場合はエラーとなってしまうので、Buttonコンポーネントでundefinedチェックをする関数を取り入れています。
こちらの関数はButtonコンポーネントでは必須となるため、あえて切り離さずコンポーネント内に記述しております。（InputやLinkのコンポーネントで実行し得るイベントも全て）

yoshin04 replied:
any型はNGです。
TypeScriptで肩を指定する意味がなくなってしまいます。
onClickで渡す関数を汎用的にするのであれば、以下のように定義することが多いです。
```
() => void
```

> onClickなどのイベントはビジネスロジック関数を実行するため、ビジネスロジック関数がundefinedの場合はエラーとなってしまうので、Buttonコンポーネントでundefinedチェックをする関数を取り入れています。

onClickが必ず渡す仕様であれば、必ず渡すようにpropsを指定しましょう。
オプショナルにする場合は、propsで渡ってきた場合にのみ実行すれば良いです。
```
type Props = {
  onClick?: () => void
}
onClick={onClick && onClick()}
```

yoshin04 commented on src/components/Common/atoms/modal-my-tip.tsx:
```diff
@@ -0,0 +1,13 @@
+import { useAuthContext } from "../../../Context/auth-context";
+
+const ModalMyTip = () => {
```
こちらも明確なつかいどころが決まっているので、atomsにはならないです。
この程度なら、コンポーネント分割する必要はないです。

もしするとしてもTextというコンポーネントを定義して、画面で使うテキストのスタイルをこのコンポーネントで管理する感じです。
```
const Text:<Props> = ({children}) => <p>{children}</p>

// 他のコンポーネントで利用する
<div className="mb-3"> 
  <Text>あなたの残高：{context?.user?.Tip}</Text>
</div>
```

yoshin04 replied:
複数カラー存在する場合は、Textコンポーネントのpropsにcolorを追加して、propsのカラーに応じてテキストカラースタイルを付与するようにコードを書くのが良いと思います。
fontWeightなども必要に応じて上記と同じように実装すると良いかと！

yoshin04 commented on src/components/Common/atoms/modal-send-button.tsx:
```diff
@@ -0,0 +1,18 @@
+import { FC } from 'react';
+import { useStateTypeShow } from '../../../types/type';
+
+const ModalSendButton:FC<{buttonValue: string, showModalItem: useStateTypeShow}> = (props) => {
```
こちらも明確なつかいどころが決まっているので、atomsにはならないです。

yoshin04 commented on src/components/Common/atoms/modal-target-name.tsx:
```diff
@@ -0,0 +1,12 @@
+import { FC } from "react";
+
+const ModalTargetName: FC<{targetName: string | undefined}> = (props) => {
```
こちらも明確なつかいどころが決まっているので、atomsにはならないです。

以下のファイルでコメントさせていただいた内容と同じですので、参考にしてみてくだだい。
src/components/Common/atoms/modal-my-tip.tsx

yoshin04 commented on src/components/Common/atoms/modal-target-tip.tsx:
```diff
@@ -0,0 +1,12 @@
+      <p className="mb-3">{props.tip}</p>
+    </div>
+  )
+}
```
こちらも明確なつかいどころが決まっているので、atomsにはならないです。

以下のファイルでコメントさせていただいた内容と同じですので、参考にしてみてくだだい。
src/components/Common/atoms/modal-my-tip.tsx

yoshin04 commented on src/components/Common/atoms/modal-wallet-button.tsx:
```diff
@@ -0,0 +1,20 @@
+import { FC } from 'react';
+import { typeSetShow, typeSetActive } from '../../../types/type';
+
+const ModalWalletButton: FC<{setShowModal: typeSetShow, setActive: typeSetActive}> = (props) => {
```
こちらも明確なつかいどころが決まっているので、atomsにはならないです。
Buttonというatomsを定義してOrgansimsから具体的な処理を注入するほうがコンポーネント設計がスッキリして、結果的にコードがわかりやすくなると思います！

yoshin04 commented on src/components/Common/atoms/title.tsx:
```diff
@@ -0,0 +1,10 @@
+  
+  return(
+    <div className="level-item">
+      <p className="has-text-dark" style={{fontSize: '4vw'}}>𝓽𝓲𝓹 𝓪𝓹𝓹𝓵𝓲𝓬𝓪𝓽𝓲𝓸𝓷</p>
```
こちらも中のテキストはchildrenにするべきですね。
そうすることで汎用性が高くなります。

yoshin04 commented on src/components/Common/atoms/other-link.tsx:
```diff
@@ -0,0 +1,21 @@
+  }
+  
+  return(
+    <div className="level-item mt-5">
```
この上マージンはリンクを使う親コンポーネントで調整するほうが良いかと！
理由は、必ず上マージンが発生してしまうので、汎用性が低くなり良いatomsとは言えなくなってしまいます。

yoshin04 commented on src/components/Common/atoms/tip.tsx:
```diff
@@ -0,0 +1,13 @@
+  return(
+    <div>
+      <p style={{fontWeight: 'bold', fontSize: '1.5vw'}}>残高：{context?.user?.Tip}</p>
+    </div>
```
こちらも明確なつかいどころが決まっているので、atomsにはならないです。

以下のファイルでコメントさせていただいた内容と同じですので、参考にしてみてくだだい。
src/components/Common/atoms/modal-my-tip.tsx

yoshin04 commented on src/components/Common/atoms/user-list-title.tsx:
```diff
@@ -0,0 +1,10 @@
+const UserListTitle = () => {
+  
+  return(
+    <div className="level-item block">
+      <p className="has-text-dark" style={{fontSize: '3.5vw'}}>ユーザ一覧</p>
+    </div>
```
こちらも明確なつかいどころが決まっているので、atomsにはならないです。

以下のファイルでコメントさせていただいた内容と同じですので、参考にしてみてくだだい。
src/components/Common/atoms/modal-my-tip.tsx

yoshin04 commented on src/components/Common/atoms/user-name.tsx:
```diff
@@ -0,0 +1,13 @@
+      <p style={{fontWeight: 'bold', fontSize: '1.5vw'}}>{context?.user?.UserName}さんようこそ!!</p>
+    </div>
+  )
+}
```
こちらも明確なつかいどころが決まっているので、atomsにはならないです。

以下のファイルでコメントさせていただいた内容と同じですので、参考にしてみてくだだい。
src/components/Common/atoms/modal-my-tip.tsx

masa-s60 committed:
[change]Atomsコンポーネントの大幅な修正

汎用性のないコンポーネントを全て汎用性のあるコンポーネント（button,input,link,text,icon-image）に変更しました。

masa-s60 committed:
[change]ログイン後のリロード時にログイン画面へ遷移してしまう動作の修正

ログイン情報のデータ保持方法をcontextからrecoilに変更し、関連するソースの修正を行いました。

masa-s60 committed:
[change, fix]関数の型定義と実行方法変更、及びコンポーネント分割の変更、及びControllerの実装修正

undefinedチェック位置を変更、関数のpropsへの渡し方を変更しました。nコンポーネント分割にOrganismsを追加し、対応コンポーネントを配置しました。nControllerの実装方法はregisterを削除、子コンポーネントであるInputのforwardRefを削除し、renderにて必要なpropsを渡す事で修正を行いました。

masa-s60 committed:
[change]コンポーネント構成の変更
コンポーネント構成を整理するため、各機能別niFeaturesを作成してコンポーネントを配置
